oauth2-test/src/main/java/org/springframework/security/test/oauth2/support/JwtAuthenticationBuilder.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright 2002-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.test.oauth2.support;
+
+import static org.springframework.security.test.oauth2.support.CollectionsSupport.putIfNotEmpty;
+
+import java.time.Instant;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.springframework.security.oauth2.jwt.Jwt;
+import org.springframework.security.oauth2.jwt.JwtClaimNames;
+import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken;
+
+/**
+ * @author Jérôme Wacongne &lt;ch4mp@c4-soft.com&gt;
+ * @since 5.2.0
+ */
+public class JwtAuthenticationBuilder<T extends JwtAuthenticationBuilder<T>> extends AbstractAuthenticationBuilder<T> {
+
+	public static final String DEFAULT_TOKEN_VALUE = "test.jwt.value";
+
+	public static final String DEFAULT_HEADER_NAME = "test-header";
+
+	public static final String DEFAULT_HEADER_VALUE = "abracadabra";
+
+	public static final Map<String, Object> DEFAULT_HEADERS =
+			Collections.singletonMap(DEFAULT_HEADER_NAME, DEFAULT_HEADER_VALUE);
+
+	private String tokenValue = DEFAULT_TOKEN_VALUE;
+
+	private Map<String, Object> headers = DEFAULT_HEADERS;
+
+	private boolean isHeadersSet = false;
+
+	/**
+	 * Solely claims will be considered at build() time. token properties that can be
+	 * stored as claims (audience, expiresAt, id, issuedAt, issuer, notBefore) are
+	 * ignored.
+	 * @param jwt fully configured JWT
+	 * @return pre-configured builder
+	 */
+	public T jwt(final Jwt jwt) {
+		final Map<String, Object> claims = new HashMap<>(jwt.getClaims());
+		if (jwt.getIssuedAt() != null) {
+			if (jwt.getClaims().containsKey(JwtClaimNames.IAT)
```

Comment:
This seems to fit better in `Jwt` itself, alongside the other validation `Jwt` does in its constructor. Do you see this validation as a better fit in the testing support?

Reply:
I dropped it from this PR, so it is possible to build incoherent tokens. In such case, claim issue and expiry dates win.

I'm a bit confused with `AbstractOAuth2Token`. Why are `issuedAt` and `expiresAt` members? It is a duplicate for corresponding claims at so many places.

I feel like a `final Map<String, Object> attributes` would be a better candidate for abstraction (don't all token have some, it sometimes being called claims?).

I wonder also if `getIssuedAt()` and `getExpiresAt()` wouldn't better be `abstract` in `AbstractOAuth2Token` and accessed from claims from each token type, using corresponding claim names.

What do you think? Should a ticket be created for that?

Reply:
Thanks for adding a ticket for this - I think it's a valuable discussion.

Reply:
I did: https://github.com/spring-projects/spring-security/issues/6807
but it was closed "won't fix" :-/